### DISCUSSION TITLE
Get symfony users by role (PostgreSQL)

### DISCUSSION THREAD

**Discussion by minato223** [symfony/symfony#57620] [Q&A]
In my Symfony application, I want to retrieve my users by role and to do this I use this logic in my repository
<img width="420" alt="Capture d’écran 2024-07-01 à 17 13 56" src="https://github.com/symfony/symfony/assets/71100950/d8398250-0d97-4e2e-bc89-ba15f26ab95a">
and I use the [martin-georgiev/postgresql-for-doctrine](https://github.com/martin-georgiev/postgresql-for-doctrine) library for the conversion function
Here my doctrine configuration 
<img width="567" alt="Capture d’écran 2024-07-01 à 17 16 46" src="https://github.com/symfony/symfony/assets/71100950/968ce239-d56e-4e5c-bcd1-7d31cd2bc387">
And here my output
<img width="1059" alt="Capture d’écran 2024-07-01 à 17 18 19" src="https://github.com/symfony/symfony/assets/71100950/f5e88cf7-8fce-4903-be2c-a529d799a52c">
When I write native SQL code, it works but in this case I need to return a QueryBuilder
<img width="588" alt="Capture d’écran 2024-07-01 à 17 19 29" src="https://github.com/symfony/symfony/assets/71100950/c9138548-776d-49d2-b4dc-9cd6b25d0874">

I work with version 7 of symfony and version 8.3.8 of php

**Comment by MatTheCat** [comment]
What if you don’t double quote `ROLE_ADMIN` in `setParameter`?

**Reply by minato223** [reply]
I got the same error, even if i remove double quote

**Reply by MatTheCat** [reply]
Ah I guess the `roles` column type is JSON and not JSONB. The library you mentioned provide a JSONB type; you should use it here.

**Reply by minato223** [reply]
I already enabled the jsonb option for the role attribute
<img width="330" alt="Capture d’écran 2024-07-02 à 22 37 23" src="https://github.com/symfony/symfony/assets/71100950/1a74cc31-4888-4cb4-a5b9-60b3d3bceccb">

**Reply by MatTheCat** [reply]
Are you sure you shouldn't be using `type` instead of `options` to configure the type?

**Reply by minato223** [reply]
I did and it works, I had to register the jsonb type in Doctrine
<img width="484" alt="Capture d’écran 2024-07-03 à 11 25 06" src="https://github.com/symfony/symfony/assets/71100950/6819b923-8ba3-4aee-9a2c-d895d8ded57b">

Thank you very much 🙏

**Comment by ker0x** [comment]
This should work:

```php
public function findByRole(string $role): mixed
{
    $rsm = $this->createResultSetMappingBuilder('u');

    $query = <<<SQL
        SELECT %s
        FROM user u
        WHERE u.roles::jsonb ?? :role
    SQL;

    $rawQuery = sprintf($query, $rsm->generateSelectClause());

    $query = $this->getEntityManager()->createNativeQuery($rawQuery, $rsm);
    $query->setParameter('role', $role);

    return $query->getResult();
}
```

**Reply by minato223** [reply]
I got this error
<img width="1036" alt="Capture d’écran 2024-07-02 à 19 39 13" src="https://github.com/symfony/symfony/assets/71100950/0030567f-b42d-44b7-926b-42ba98af7f45">

**Comment by minato223** [comment]
It's been two nights that I haven't slept because of that so if beyond that you have other methods to recover users by their role, I'm interested
